src/dune_engine/build_system.ml

```diff
@@ -476,10 +465,24 @@ end = struct
           in
           let produced_targets =
             match sandbox with
-            | None ->
+            | None -> (
               (* Directory targets are not allowed for non-sandboxed actions, so
                  the call below should not raise. *)
-              Targets.Produced.of_validated_files_exn targets
+              match Targets.Produced.of_validated targets with
+              | Ok t -> t
+              | Error (`Directory dir, (Unix.ENOENT, _, _)) ->
```

Comment:
I think I'd prefer to have this error-handling logic in `Targets.Produced.of_validated_exn` to keep `build_system.ml` as simple as possible.

Reply:
Good idea. I'm still a little unsatisfied that `Sandbox` also has some of these checks. I'll try to unify things somehow. Although if it proves to be a lot of work, I'll move this to a separate PR.